### PR TITLE
Use bisq's bitcoinj 0.15.8 - september 2020

### PR DESCRIPTION
**Pull request by oscarguindzberg**
Adapt Bisq to bitcoinj 0.15.8  (see https://github.com/bisq-network/bitcoinj/commits/bisq_0.15.8)
That is a pre-requisite for segwit support.
This PR replaces https://github.com/bisq-network/bisq/pull/4270

### CODE REVIEW

**Comment by oscarguindzberg** [comment]
I see codacy suggested a couple of changes, I will fix them ASAP

File: core/src/main/java/bisq/core/btc/setup/BisqKeyChainGroupStructure.java
Comment (chimp1984) [inline core/src/main/java/bisq/core/btc/setup/BisqKeyChainGroupStructure.java:62]: Aren't there more script types (MultiSig) to check?

File: core/src/main/java/bisq/core/btc/setup/WalletsSetup.java
Comment (chimp1984) [inline core/src/main/java/bisq/core/btc/setup/WalletsSetup.java]: Are httpSeed still used by default in BitcoinJ? If so I would prefer to still make sure to not use http seeds as they have some privacy issues.

File: core/src/main/java/bisq/core/btc/setup/WalletConfig.java
Comment (chimp1984) [inline core/src/main/java/bisq/core/btc/setup/WalletConfig.java]: I guess the else case should be P2WPKH

**Review by chimp1984 — CHANGES_REQUESTED** [review]

**Comment by chimp1984** [comment]
@ripcurlx why it got flagged incorrectly (altcoin)?

**Comment by ripcurlx** [comment]
> @ripcurlx why it got flagged incorrectly (altcoin)?

It touches the assets module.

File: core/src/main/java/bisq/core/btc/setup/BisqKeyChainGroupStructure.java
Comment (oscarguindzberg) [reply core/src/main/java/bisq/core/btc/setup/BisqKeyChainGroupStructure.java:62]: Even though Bisq uses multisig (eg for payout tx signing), bisq does not use a multisig keychain.
A multisig keychain (MarriedKeyChain in bitcoinj) makes sense when you want to repeatedly generate new multisig addresses for the same group of people.

File: core/src/main/java/bisq/core/btc/setup/WalletConfig.java
Comment (oscarguindzberg) [reply core/src/main/java/bisq/core/btc/setup/WalletConfig.java]: That will be true when we implement segwit.
For now, this is just migration to bitcoinj 0.15
The code is a placeholder for the code that is coming soon (see the comment in the previous line).
I will update the code to:
Script.ScriptType preferredOutputScriptType = Script.ScriptType.P2PKH;
To avoid confusions until segwit is implemented.

File: core/src/main/java/bisq/core/btc/setup/WalletsSetup.java
Comment (oscarguindzberg) [reply core/src/main/java/bisq/core/btc/setup/WalletsSetup.java]: http seeds are not used by default in bitcoinj and are not used in bisq as is.
Having said that, I had another look at this and found they might be used if a bisq dev activates a certain feature of bitcoinj that is not being used today. So, I will add PeerGroup.setIgnoreHttpSeeds(true) again.

File: core/src/main/java/bisq/core/btc/setup/BisqKeyChainGroupStructure.java
Comment (chimp1984) [reply core/src/main/java/bisq/core/btc/setup/BisqKeyChainGroupStructure.java:62]: Ok, thanks. Did not see the full context.

File: core/src/main/java/bisq/core/btc/setup/WalletConfig.java
Comment (chimp1984) [reply core/src/main/java/bisq/core/btc/setup/WalletConfig.java]: Ah missed the comment.

**Comment by oscarguindzberg** [comment]
I fixed codacy suggestions, but 2 of them. They say I should avoid calling toString() on a String object, but actually Sha256Hash.toString() is being called.

**Comment by chimp1984** [comment]
> I fixed codacy suggestions, but 2 of them. They say I should avoid calling toString() on a String object, but actually Sha256Hash.toString() is being called.

Its in TransactionsListItem.java trade.getDepositTx().getTxId().toString()

**Review by chimp1984 — DISMISSED** [review]
preliminary utACK
But I would like to do a more in depth review in a few days. So better wait for after release with merge and another ACK as its high risk area with wallet code changes.

**Comment by ripcurlx** [comment]
@oscarguindzberg Could you please have a look at the two unnecessary `toString` calls Codacy is complaining about, so this PR is mergeable? Thanks!

**Review by chimp1984 — DISMISSED** [review]
utACK

File: core/src/main/java/bisq/core/btc/model/AddressEntryList.java
Comment (sqrrm) [inline core/src/main/java/bisq/core/btc/model/AddressEntryList.java]: Prefer to break lines at 120 characters.

File: gradle/witness/gradle-witness.gradle
Comment (sqrrm) [inline gradle/witness/gradle-witness.gradle:71]: Upgraded btcd-cli4j, why was this necessary?

**Review by sqrrm — COMMENTED** [review]
Tested in regtest and wallet conversion seems to have worked fine. Sending coins work fine, BSQ parsing looks correct.

What I haven't done is completely understood the changes to WalletConfig.java. The rest looks ok to me.

**Review by chimp1984 — DISMISSED** [review]
utACK

**Comment by oscarguindzberg** [comment]
@ripcurlx the two "unnecessary toString() calls Codacy is complaining about" seem like a Codacy bug to me. toString() is being called on Sha256Hash instances, not on String instances.

File: gradle/witness/gradle-witness.gradle
Comment (oscarguindzberg) [reply gradle/witness/gradle-witness.gradle:71]: Kind of long story...
A year ago I upgraded some Bisq dependencies to match bitcoinj 0.15.1 dependencies. eg upgraded guava from 20 to 27.
Then, when bitcoinj 0.15.8 came out with guava 28.2 , I upgraded Bisq to use that guava version.
None of those changes were merged yet, we are merging it as part of this PR.

Some time ago this code was merged into the btcd-cli4j repo https://github.com/bisq-network/btcd-cli4j/pull/4 . And then you reverted part of the change https://github.com/bisq-network/btcd-cli4j/pull/8.

So, I am not actually upgrading btcd-cli4j but downgrading it to the version that declares a dependency on guava 27.

In any case, bisq's gradle resolves not to include the guava version declared in btcd-cli4j but the version declared on its own build.gradle file.

To make things clear a new btcd-cli4j version declaring a dependency to guava 28.2-jre could be created, then we could upgrade to that btcd-cli4j version in bisq.

**Comment by oscarguindzberg** [comment]
@sqrrm @ripcurlx I think this PR is ready for merging. Please, let me know if there is something I did not answer/fix.

File: core/src/main/java/bisq/core/btc/model/AddressEntryList.java
Comment (oscarguindzberg) [reply core/src/main/java/bisq/core/btc/model/AddressEntryList.java]: done

**Comment by oscarguindzberg** [comment]
@sqrrm  I re-requested review from you, I guess that is necessary to merge the PR since I pushed 3 new commits with a bugfix and code styling changes.

**Review by sqrrm — DISMISSED** [review]
ACK

Waiting for @ripcurlx to test this before merging as well. Also need a resolution to the codacy complaints.

**Review by chimp1984 — DISMISSED** [review]
utACK

**Comment by chimp1984** [comment]
@sqrrm @ripcurlx The codady complains are false positives. getTxId() returns a ShaHash and toString() is correct on that. Seems a bug in codacy. Please ignore that and merge or deactivate that rule.

**Comment by sqrrm** [comment]
Impossible to ignore, it blocks merging. So either we're changing codacy rules or we write bad code to get around codacy. It's a pity that's it's so annoying.

**Comment by oscarguindzberg** [comment]
@sqrrm I added a commit https://github.com/bisq-network/bisq/pull/4504/commits/83a1e2f044901e773cd848660f4a08da2ce09e39 to make codacy happy. Check passes now.

**Review by sqrrm — DISMISSED** [review]

**Comment by oscarguindzberg** [comment]
I just signed my commits. I had to push again. I guess the PR requires approval again.

**Review by sqrrm — APPROVED** [review]

**Comment by chimp1984** [comment]
> Impossible to ignore, it blocks merging. So either we're changing codacy rules or we write bad code to get around codacy. It's a pity that's it's so annoying.

Thats the real dnager with that too strict and inflexible robot. We bow down and do things we would not do otherwise (for instance i use less exception handling now as it enforces me to write custom exceptions which is often boilerplate, leading at the end to less safe code). We are the masters not codacy.

**Review by chimp1984 — CHANGES_REQUESTED** [review]
NACK (to last commit bowing down to stupid codacy bot)

I think this is really wrong that we start to write less readible code because of that stupid bot. @ripcurlx Please deactivate codacy until we find a more flexible solution. I find this inacceptable that we cannot override clear mistakes or bad settings in codacy.

**Comment by oscarguindzberg** [comment]
@chimp1984 The last commit (https://github.com/bisq-network/bisq/commit/83a1e2f044901e773cd848660f4a08da2ce09e39) does not seem wrong or less readable to me, it is just another way to do the same thing. Please, please, don't NACK this PR... it was so hard to be ready to merge. Let's fix codacy in another issue/PR.

**Review by chimp1984 — APPROVED** [review]
utACK

**Comment by chimp1984** [comment]
Ok, I just got too  many times annoyed by a tool which seems to makes us slaves.

**Comment by cd2357** [comment]
> Impossible to ignore, it blocks merging. So either we're changing codacy rules or we write bad code to get around codacy. It's a pity that's it's so annoying.

@sqrrm are you sure?

I played around with Codacy and it can be ignored when merging:

github-repo-merge-despite-codacy

**Comment by sqrrm** [comment]
I guess admins can override codacy, but I'm not an admin so I can't.

**Comment by cd2357** [comment]
Hmm.

Alternatively, Codacy can be removed from the "branch protection" settings.

github-codacy-branch-protection

In that case, the Codacy checks are only informative and merging is possible without admin permissions.

github-codacy-branch-protection-result

Which seems like a sane setup.

**Comment by cd2357** [comment]
Other checks can still be enforced though, like successful travis builds.

**Comment by sqrrm** [comment]
@cd2357 making them recommendations rather than enforced sounds like the more flexible and sane thing to do, all things considered. Thanks for looking that up, @ripcurlx could you make codacy optional? I think it could save some headaches.

**Pull request merged**
